### ISSUE TITLE
[Feature Request] Parallel quorum write

### ISSUE THREAD

**Issue by vasyaabr** [ClickHouse/ClickHouse#3950]
Trying to INSERT data from multiple writers to Replicated... table with insert_quorum>1 usually results in exception: 

> DB::Exception: Quorum for previous write has not been satisfied yet.

It would be great to implement some sort of queue (or other mechanics), which will allow multiple INSERTs in parallel with enabled quorum write.

**Comment by kaijianding** [comment]
echo this.
I found the same error when I set insert_quorum=2, these 2 exceptions are raised in version: 19.7.3.1:

Another quorum insert has been already started
Quorum for previous write has not been satisfied yet

**Comment by vasyaabr** [comment]
I will up this issue

**Comment by alexey-milovidov** [comment]
This task is assigned to external contributor Alexandra Latysheva.

**Comment by haoch** [comment]
Hi @alexey-milovidov @alexelex, any update on this issue? 
Looks like clickhouse `insert_quorum` works with some kind of `exclusive lock` with zookeep path `storage.zookeeper_path + "/quorum/status" ` (`/clickhouse/tables/${SHARD}/${TABLE}/quorum/status`). Why is it necessary?  Or even so, can clickhouse block the parallel inert request in case if the lock taken to avoid throwing exception with `DB::Exception: Quorum for previous write has not been satisfied yet.`?

**Comment by alexey-milovidov** [comment]
The task should be not hard to implement, but it is blocked by another task:

Don't ignore data from duplicate INSERTS that were sent to replicas.
So, if user send the same data to several replicas manually, this data should be accepted as if it was received by replication protocol.

This task is unassigned, but I'm currently looking at the ways of implementation.

**Comment by filimonov** [comment]
https://github.com/ClickHouse/ClickHouse/pull/15601

**Comment by filimonov** [comment]
Implemented in #15601, available since 20.10. 

Can be enabled with `insert_quorum_parallel=1`. 

The feature in its current form has a drawback - it can't work together with the sequential consistency model, obviously 'sequential' is problematic when you insert in parallel. 

That also means that replicas can see the parts even if a quorum were not satisfied for them (insert is still running, or was reported as failed). 

So that actually just allows sending feedback about unsatisfied insert to the inserter, keeping the usual eventual consistency model in place (at a certain moment of time the problematic replica will download the part needed to satisfy the quorum).
